include/deal.II/grid/manifold_lib.h

```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+ * setting only a curved manifold on the boundary of a mesh, since the latter
+ * will need to switch from a curved description to a straight description in a
+ * single layer of elements, which causes an error order on cells close to the
+ * boundary that does not exceed 3 no matter how high the degree of the
+ * polynomial mapping and the finite element space. Using this class instead,
+ * the curved nature of the manifold that is originally contained in one
+ * <i>coarse</i> mesh layer will be applied to more than one <i>fine</i> mesh
+ * layer once the mesh gets refined, restoring the optimal convergence rates of
+ * the underlying finite element and mapping as appropriate.
+ *
+ * If no curved boundaries surround a coarse cell, this class reduces to a flat
+ * manifold description.
+ *
+ * To give an example of using this class, the following code attaches a
+ * transfinite manifold to a circle:
+ *
+ * @code
+ * PolarManifold<dim> polar_manifold;
+ * TransfiniteInterpolationManifold<dim> inner_manifold;
+ *
+ * Triangulation<dim> triangulation;
+ * GridGenerator::hyper_ball (triangulation);
+ *
+ * triangulation.set_all_manifold_ids(1);
+ * triangulation.set_all_manifold_ids_on_boundary(0);
+ * triangulation.set_manifold (0, polar_manifold);
+ * inner_manifold.initialize(triangulation);
+ * triangulation.set_manifold (1, inner_manifold);
+ * triangulation.refine_global(4);
+ * @endcode
+ *
+ * In this code, we first set all manifold ids to the id of the transfinite
+ * interpolation, and then re-set the manifold ids on the boundary to identify
+ * the curved boundary described by the polar manifold. With this code, one
+ * gets a really nice mesh:
+ *
+ * <p ALIGN="center">
+ * @image html circular_mesh_transfinite_interpolation.png
+ * </p>
+ *
+ * which is obviously much nicer than the polar manifold applied to just the
+ * boundary:
+ *
+ * <p ALIGN="center">
+ * @image html circular_mesh_only_boundary_manifold.png
+ * </p>
+ *
+ * @note In the implementation of this class, the manifolds surrounding a
+ * coarse cell are queried repeatedly to compute points on their interior. For
+ * optimal mesh quality, those manifolds should be compatible with a chart
+ * notion. For example, computing a point that is 0.25 along the line between
+ * two vertices using the weights 0.25 and 0.75 for the two vertices should
+ * give the same result as first computing the mid point at 0.5 and then again
+ * compute the midpoint between the first vertex and coarse mid point. This is
+ * the case for PolarManifold but not for Spherical manifold, so be careful
+ * when using the latter. In case the quality of the manifold is not good
+ * enough, upon mesh refinement it may happen that the transformation to a
+ * chart inside the get_new_point() or add_new_point() methods produces points
+ * that are outside the unit cells, then this class throws an exception of type
+ * Manifold@<dim,spacedim@>::ExcTransformationFailed. If that happens, the mesh
```

Comment:
I must also add that it took me a few days to get this robust and re-wrote the logic two times because one really needs to try the inverse mapping with the given points, it might fail, and one needs to try with another set (usually a neighbor).